File: src/lib/api/payments.ts
import { invoke } from "@tauri-apps/api/core"
import type { Order } from "@/lib/api/orders"

export type DiscountType = "PERCENT" | "AMOUNT"
export type DiscountScope = "ORDER" | "ITEM"
export type PaymentMethod = "CASH" | "QR"

export type Discount = {
  id: number
  name: string
  type: DiscountType
  value: number
  scope: DiscountScope
  is_active: boolean
  valid_from: string | null
  valid_to: string | null
  sort_order: number
}

export type DiscountInput = {
  name: string
  type: DiscountType
  value: number
  scope: DiscountScope
  valid_from: string | null
  valid_to: string | null
  sort_order: number
}

export type OrderDiscount = {
  id: number
  order_id: number
  discount_id: number | null
  name: string
  type: DiscountType
  value: number
  amount_applied: number
}

export type ApplyDiscountInput = {
  discount_id: number | null
  name: string
  type: DiscountType
  value: number
}

export type Payment = {
  id: number
  order_id: number
  method: PaymentMethod
  amount: number
  tendered: number | null
  change_due: number | null
  paid_at: string
  ref_note: string | null
}

export type PaymentInput = {
  method: PaymentMethod
  amount: number
  tendered: number | null
  ref_note: string | null
}

export const listDiscounts = (includeInactive = true) =>
  invoke<Discount[]>("list_discounts", { includeInactive })

export const createDiscount = (payload: DiscountInput) =>
  invoke<Discount>("create_discount", { payload })

export const updateDiscount = (id: number, payload: DiscountInput) =>
  invoke<Discount>("update_discount", { id, payload })

export const setDiscountActive = (id: number, isActive: boolean) =>
  invoke<void>("set_discount_active", { id, isActive })

export const applyDiscount = (orderId: number, payload: ApplyDiscountInput) =>
  invoke<Order>("apply_discount", { orderId, payload })

export const removeOrderDiscount = (orderDiscountId: number) =>
  invoke<Order>("remove_order_discount", { orderDiscountId })

export const addPayment = (orderId: number, payload: PaymentInput) =>
  invoke<Order>("add_payment", { orderId, payload })

export const finalizeOrder = (orderId: number) => invoke<Order>("finalize_order", { orderId })

export const generateBillHtml = (orderId: number) =>
  invoke<string>("generate_bill_html", { orderId })

export const generateBillPdf = (orderId: number) => invoke<string>("generate_bill_pdf", { orderId })

File: src-tauri/src/domain/payments.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum DiscountType {
    Percent,
    Amount,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum DiscountScope {
    Order,
    Item,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum PaymentMethod {
    Cash,
    Qr,
}

#[derive(Debug, Clone, Serialize)]
pub struct Discount {
    pub id: i64,
    pub name: String,
    pub r#type: DiscountType,
    pub value: i64,
    pub scope: DiscountScope,
    pub is_active: bool,
    pub valid_from: Option<String>,
    pub valid_to: Option<String>,
    pub sort_order: i64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct DiscountInput {
    pub name: String,
    pub r#type: DiscountType,
    pub value: i64,
    pub scope: DiscountScope,
    pub valid_from: Option<String>,
    pub valid_to: Option<String>,
    pub sort_order: i64,
}

#[derive(Debug, Clone, Serialize)]
pub struct OrderDiscount {
    pub id: i64,
    pub order_id: i64,
    pub discount_id: Option<i64>,
    pub name: String,
    pub r#type: DiscountType,
    pub value: i64,
    pub amount_applied: i64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ApplyDiscountInput {
    pub discount_id: Option<i64>,
    pub name: String,
    pub r#type: DiscountType,
    pub value: i64,
}

#[derive(Debug, Clone, Serialize)]
pub struct Payment {
    pub id: i64,
    pub order_id: i64,
    pub method: PaymentMethod,
    pub amount: i64,
    pub tendered: Option<i64>,
    pub change_due: Option<i64>,
    pub paid_at: String,
    pub ref_note: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct PaymentInput {
    pub method: PaymentMethod,
    pub amount: i64,
    pub tendered: Option<i64>,
    pub ref_note: Option<String>,
}
